File: src/main/scala/net/michalsitko/HostLevel.scala
package net.michalsitko

import java.net.InetSocketAddress

import akka.NotUsed
import akka.actor.ActorSystem
import akka.http.scaladsl.Http.HostConnectionPool
import akka.http.scaladsl.model.{HttpRequest, HttpResponse}
import akka.http.scaladsl.settings.{ClientConnectionSettings, ConnectionPoolSettings}
import akka.http.scaladsl.{ClientTransport, Http}
import akka.stream.ActorMaterializer
import akka.stream.scaladsl.{Flow, Sink, Source}
import net.michalsitko.utils.ResultLogger

import scala.concurrent.Future
import scala.util.{Failure, Success, Try}

object HostLevel extends AnyRef with ResultLogger {
  def main(args: Array[String]): Unit = {
    implicit val system = ActorSystem()
    implicit val materializer = ActorMaterializer()
    implicit val ec = system.dispatcher

    val source: Source[(HttpRequest, String), NotUsed] = Source(List("/")).map(path => (HttpRequest(uri = path), path))

    val proxySettings = new InetSocketAddress("localhost", 8888)
    val transport = ClientTransport.proxy(None, proxySettings, ClientConnectionSettings(system))

    val connectionPoolSettings = ConnectionPoolSettings(system).withTransport(transport)
    val poolClientFlow: Flow[(HttpRequest, String), (Try[HttpResponse], String), HostConnectionPool] =
      Http().cachedHostConnectionPool[String]("www.scala-lang.org", 80, connectionPoolSettings)

    val result: Future[(Try[HttpResponse], String)] = source.via(poolClientFlow).runWith(Sink.head)

    result
      .flatMap {
        case (Success(response), _) =>
          logResult(Future.successful(response))
        case (Failure(ex), _) =>
          Future.successful(())
      }
      .andThen {
        case _ =>
          Http().shutdownAllConnectionPools()
          materializer.shutdown()
          system.terminate()
      }
  }
}

File: src/main/java/net/michalsitko/java/HostLevel.java
package net.michalsitko.java;

import akka.NotUsed;
import akka.actor.ActorSystem;
import akka.http.javadsl.ClientTransport;
import akka.http.javadsl.ConnectHttp;
import akka.http.javadsl.HostConnectionPool;
import akka.http.javadsl.Http;
import akka.http.javadsl.model.*;
import akka.http.javadsl.settings.ClientConnectionSettings;
import akka.http.javadsl.settings.ConnectionPoolSettings;
import akka.japi.Pair;
import akka.stream.ActorMaterializer;
import akka.stream.Materializer;
import akka.stream.javadsl.Flow;
import akka.stream.javadsl.Sink;
import akka.stream.javadsl.Source;
import akka.util.ByteString;

import scala.util.Try;

import java.net.InetSocketAddress;
import java.util.Optional;
import java.util.concurrent.CompletionStage;
import java.util.function.Function;

class HostLevelClient {

    private final ActorSystem system;
    private final Materializer materializer;
    private final Flow<Pair<HttpRequest, Integer>, Pair<Try<HttpResponse>, Integer>, HostConnectionPool> poolClientFlow;

    public HostLevelClient(ActorSystem sys, Materializer mat) {
        system = sys;
        materializer = mat;

        InetSocketAddress addr = new InetSocketAddress("localhost", 8888);
        ClientTransport clientTransport = ClientTransport.proxy(Optional.empty(), addr, ClientConnectionSettings.create(system));
        ConnectionPoolSettings settings = ConnectionPoolSettings.create(system).withTransport(clientTransport);
        poolClientFlow = Http.get(system).cachedHostConnectionPool(
                ConnectHttp.toHost("www.scala-lang.org", 443), settings, system.log(), materializer);
    }

    public ActorSystem getSystem() {
        return system;
    }

    public Materializer getMaterializer() {
        return materializer;
    }

    public void close() {
        System.out.println("Shutting down client");
        Http.get(system).shutdownAllConnectionPools().whenComplete((s, f) -> system.terminate());
    }

    private Uri uri = Uri.create("https://www.scala-lang.org/resources/css/prettify.css");

    public <U> CompletionStage<U> hostLevel(Function<Pair<Try<HttpResponse>, Integer>, CompletionStage<U>> responseHandler) {
        return Source.single(Pair.create(HttpRequest.create().withUri(uri), 42))
                .via(poolClientFlow)
                .runWith(Sink.head(), materializer)
                .thenComposeAsync(responseHandler);
    }

}

public class HostLevel {
    public static void main(String[] args) {
        System.out.println("Starting...");
        final ActorSystem system = ActorSystem.create();
        final Materializer materializer = ActorMaterializer.create(system);

        HostLevelClient client = new HostLevelClient(system, materializer);
        client.hostLevel(success ->
                success.first().get().entity()
                        .getDataBytes()
                        .runFold(ByteString.empty(), ByteString::concat, client.getMaterializer())
                        .handle((byteString, f) -> {
                            if (f != null) {
                                System.err.println("Failure when handle the HttpEntity: " + f.getMessage());
                            } else {
                                System.out.println("Result: " + byteString.utf8String());
                            }
                            return NotUsed.getInstance();
                        })
        ).whenComplete((success, throwable) -> client.close());
    }
}

